### ISSUE TITLE
Empty systems still exist in cubical agda

### ISSUE THREAD

**Issue by 3abc** [agda/agda#3583]
After https://github.com/agda/agda/issues/3415 is closed we are supposedly not going to see empty system?

However consider this example:

```agda
test : (A : Set) (x y : A) (p : x ≡ y) → Set
test A x y p = { }0
  where
  u : (i j k : I) → Partial (i ∧ j) A
  u i j k (i = i1) (j = i1) = p k
  f : I → I → A
  f i j = hcomp (u i j) x
```

Normalize `f i0 i0` in that hole I get `hcomp (λ i → empty) x`.

For your reference
```
$ agda --version
Agda version 2.6.0-1becd2e
```
it's the 2nd latest commit.

**Comment by mortberg** [comment]
This is the expected behavior. As discussed in the PR we need to impose a "validity" constraint on systems in order to never get any empty ones. The validity constraint says that a system has to be specified by a classical tautology. In your example the formula is `i /\ j` which is not a tautology and the system is hence invalid. The invariant that we do have is that if we start with a valid system then we can never end up with an empty system (this follows from the way transp/hcomp is implemented internally).

We should maybe introduce a validity check to the typechecker when checking hcomp's so that your example is rejected, but I'm not really sure if this is worth it. No matter what I don't think it has to be done for 2.6.0

**Comment by 3abc** [comment]
> The validity constraint says that a system has to be specified by a classical tautology.

Does this constraint mean that, if we think of the system as in a Boolean algebra instead of a de Morgan algebra, it always evaluate to `i1`? i.e. `i ∨ ~ i` or `i ∨ ~ i ∨ j` is valid and `i ∧ j` is not? So if at least two sides of a cube is included when we do `hcomp` then there will be no empty systems.

**Comment by mortberg** [comment]
> > The validity constraint says that a system has to be specified by a classical tautology.
> 
> Does this constraint mean that, if we think of the system as in a Boolean algebra instead of a de Morgan algebra, it always evaluate to `i1`? i.e. `i ∨ ~ i` or `i ∨ ~ i ∨ j` is valid and `i ∧ j` is not? So if at least two sides of a cube is included when we do `hcomp` then there will be no empty systems.

Exactly! The inspiration comes from Def 12 of https://www.cs.cmu.edu/~cangiuli/papers/ccctt.pdf

**Comment by 3abc** [comment]
> We should maybe introduce a validity check to the typechecker when checking hcomp's so that your example is rejected, but I'm not really sure if this is worth it.

I feel like such a check will make `hfill`, `comp` and `fill` harder to define? I'm not sure if the systems in their current definition are guaranteed valid. The only thing I can think of is currently
`comp (λ i → p i) (λ _ → empty) (inc a)` is the same as `hcomp (λ i → empty) (transp (λ i → p i) i0 a)`, not sure how this will work out if a validity check is imposed.

**Comment by Saizan** [comment]
@3abc You can replace uses of `hcomp` with `ghcomp` if your system is not known to be valid. as described in #3415.

**Comment by 3abc** [comment]
> @3abc You can replace uses of `hcomp` with `ghcomp` if your system is not known to be valid. as described in #3415.

Is `ghcomp` already available and can be imported? 

Or do I have to define it as (I'm not sure if this is the best way)

```Agda
ghcomp : ∀ {ℓ} {A : Set ℓ} {φ : I}
           (u : I → .(IsOne φ) → A)
           (a : Sub A φ (u i0)) → A
ghcomp {φ = φ} u a = hcomp {φ = φ ∨ ~ φ}
                           (λ j → λ { (φ = i1) → u j 1=1
                                    ; (φ = i0) → ouc a} )
                           (ouc a)
```

**Comment by Saizan** [comment]
@3abc I don't think it's defined anywhere. Your definition is fine except the type of `u` would be better as `I → Partial φ A`
